src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+#nullable enable
+namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.AddExplicitCast
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.AddExplicitCast), Shared]
+    [ExtensionOrder(After = PredefinedCodeFixProviderNames.ChangeToYield, Before = PredefinedCodeFixProviderNames.GenerateConstructor)]
+    internal partial class AddExplicitCastCodeFixProvider : SyntaxEditorBasedCodeFixProvider
+    {
+        /// <summary>
+        /// CS0266: Cannot implicitly convert from type 'x' to 'y'. An explicit conversion exists (are you missing a cast?)
+        /// </summary>
+        private const string CS0266 = nameof(CS0266);
+
+        /// <summary>
+        /// CS1503: Argument 1: cannot convert from 'double' to 'int'
+        /// </summary>
+        private const string CS1503 = nameof(CS1503);
+        private const int MaximumConversionOptions = 3;
+
+        [ImportingConstructor]
+        public AddExplicitCastCodeFixProvider()
+        {
+        }
+
+        public sealed override ImmutableArray<string> FixableDiagnosticIds
+        {
+            get => ImmutableArray.Create(CS0266, CS1503);
+        }
+
+        internal sealed override CodeFixCategory CodeFixCategory => CodeFixCategory.CodeStyle;
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var diagnostic = context.Diagnostics.First();
+
+            var root = await document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            if (root != null && syntaxFacts != null && semanticModel != null &&
+                TryGetTargetNode(root, diagnostic.Location.SourceSpan) is ExpressionSyntax targetNode)
+            {
+                var exactSolution = GetTypeInfo(semanticModel, root, targetNode, cancellationToken, out var nodeType, out var conversionType, out var potentialConvTypes);
```

Comment:
Semantics of these method seem a bit strange. I think it might be best to make following changes:

1. Rename the method to `TryGetTargetTypeInfo`, which returns `false` when no potential conversion types are available, and `true` when one or more conversion types are available.
2. The call site here is changed to:
   1. Check for `true` return value, and assert that `potentialConvTypes > 0` when `true`.
   1. If `potentialConvTypes == 1`, offer `Add_explicit_cast`, i.e. the logic inside this `if` part.
   2. Else iterate through `potentialConvTypes` and offer a code fix for each potential type, i.e. the logic inside your `else if` part below.

Reply:
I would also remove the parameter `conversionType` and handle the case with single conversion type by just having `potentialConvTypes` return an array with just one element.